### PR TITLE
bpo-40412: Nullify inittab_copy during finalization

### PR DESCRIPTION
**Pull request by indygreg**
Otherwise we leave a dangling pointer to free'd memory. If we
then initialize a new interpreter in the same process and call
PyImport_ExtendInittab, we will (likely) crash when calling
PyMem_RawRealloc(inittab_copy, ...) since the pointer address
is bogus.

https://bugs.python.org/issue40412

Automerge-Triggered-By: @brettcannon

### CODE REVIEW

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

Good catch!  How did you notice this?

**Comment by indygreg** [comment]
I discovered it while writing tests for PyOxidizer's custom meta path importer. The tests initialize several Python interpreters per process.

Also, is it too late to get this patch in 3.8.3?

**Comment by brettcannon** [comment]
@indygreg mind doing a quick news entry? This close to beta means people are starting to care about every change more and more. You can use https://blurb-it.herokuapp.com/ as linked from "Details" for the bedevere/news line if you want to use a web form to fill it out.

File: Misc/NEWS.d/next/C API/2020-05-01-17-28-04.bpo-40412.dE0D8N.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Nullify inittab_copy during finalization. Before, Python interpreter finalization would leave a pointer reference to released memory. If another interpreter were initialized in the same process, it would attempt to access this memory, likely leading to a crash. This was only reproducible if Python were embedded in a larger application and does not affect any application distributed by Python itself. Patch by Gregory Szorc.
```
Comment (brettcannon) [inline Misc/NEWS.d/next/C API/2020-05-01-17-28-04.bpo-40412.dE0D8N.rst]:
```suggestion
Nullify inittab_copy during finalization, preventing future interpreter initializations in an embedded situation from crashing. Patch by Gregory Szorc.
```

**Review by brettcannon — CHANGES_REQUESTED** [review]
Simplification of the news entry and then this is ready to go!

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
OK, thanks to the flag being set for submitting my own suggestions, this PR looks good to go! I have set it to automerge and backport to 3.8 and 3.7.

**Comment by brettcannon** [comment]
I also tossed in a 3.6 backport in case Ned wants to take it due to it being a memory thing.

**Comment by indygreg** [comment]
While NULLifying the invalid pointer is certainly worthy of backporting everywhere, I'm not sure this issue is reproducible on Python <3.8. The reason is that the code that clears this pointer appears to be called after `Py_FinalizeEx()`. I'm unsure if this code existed before initialization API refactors in 3.8. @vstinner would know.

I've already thrown away PyOxidizer's code to support Python 3.7 and can't easily test. This will require a source inspection to validate against older versions.

**Comment by vstinner** [comment]
Hum, inittab_copy is freed by pymain_free(). I added this function to clear objects and memory which **cannot** be freed by Py_Finalize() for backward compatibility.

The [PEP 587](https://www.python.org/dev/peps/pep-0587/) says:

> PyImport_FrozenModules, PyImport_AppendInittab() and PyImport_ExtendInittab() functions are still relevant and continue to work as previously. They should be set or called after Python preinitialization and before the Python initialization.

I understood that PyImport_AppendInittab() can be called only once, but Py_Initialize() and Py_Finalized() can be called multiple times: the PyImport_AppendInittab() call is expected to be saved even after Py_Finalized().

I understand that this change only impacts C code which calls Py_RunMain() or Py_Main() multiple times. I'm not sure if it was supported previously, but it's a good thing if slowly Python supports it :-)

**Comment by vstinner** [comment]
> I'm not sure this issue is reproducible on Python <3.8

I introduced the issue in Python 3.7 with commit 92a3c6f493ad411e4cf0acdf305ef4876aa90669:

```
commit 92a3c6f493ad411e4cf0acdf305ef4876aa90669
Author: Victor Stinner <victor.stinner@gmail.com>
Date:   Wed Dec 6 18:12:59 2017 +0100

    bpo-32030: Add _PyImport_Fini2() (#4737)
    
    PyImport_ExtendInittab() now uses PyMem_RawRealloc() rather than
    PyMem_Realloc(). PyImport_ExtendInittab() can be called before
    Py_Initialize() whereas only the PyMem_Raw allocator is supposed to
    be used before Py_Initialize().
    
    Add _PyImport_Fini2() to release the memory allocated by
    PyImport_ExtendInittab() at exit. PyImport_ExtendInittab() now forces
    the usage of the default raw allocator, to be able to release memory
    in _PyImport_Fini2().
```

A backport of the fix sounds safe, but I don't think that it is safe to call Py_Main() multiple times in Python 3.7. Maybe it works, I don't know :-)

**Pull request merged**